File: app/ios/Wallet/Sources/Logging/DatadogSpanImpl.swift
import DatadogTrace
import Shared

class DatadogSpanImpl: DatadogSpan {
    var resourceName: String? {
        didSet {
            if let name = resourceName {
                span.setOperationName(name)
            }
        }
    }

    let span: OTSpan

    init(span: OTSpan) {
        self.span = span
    }

    func setTag(key: String, value: String) {
        span.setTag(key: key, value: value)
    }

    func finish() {
        span.finish()
    }

    func finish(cause: KotlinThrowable) {
        span.setError(
            kind: cause.description,
            message: cause.message ?? "",
            stack: KotlinArrayIterator(cause.getStackTrace()).map { $0 as String }
                .joined(separator: "\n"),
            file: ""
        )

        span.finish()
    }
}

private class KotlinArrayIterator<T: AnyObject>: Sequence, IteratorProtocol {
    typealias Element = T

    let inner: KotlinArray<T>
    var index: Int32 = 0

    init(_ array: KotlinArray<T>) {
        inner = array
    }

    func next() -> T? {
        guard index < inner.size else {
            return nil
        }

        let result = inner.get(index: index)
        index = index + 1
        return result
    }
}

File: app/libs/datadog/impl/src/androidMain/kotlin/bitkey/datadog/DatadogSpanImpl.kt
package bitkey.datadog

import com.datadog.opentracing.DDSpan
import io.opentracing.Span

internal data class DatadogSpanImpl(
  val span: Span,
) : DatadogSpan {
  override var resourceName: String? = null
    set(value) {
      (span as DDSpan).resourceName = value
      field = value
    }

  override fun setTag(
    key: String,
    value: String,
  ) {
    span.setTag(key, value)
  }

  override fun finish() {
    span.finish()
  }

  override fun finish(cause: Throwable) {
    (span as DDSpan).setError(true)
    setTag("error.type", cause::class.simpleName.orEmpty())
    setTag("error.msg", cause.message.orEmpty())
    setTag("error.stack", cause.stackTraceToString())
    span.finish()
  }
}
